File: challenges/rate-limiter/typescript/src/rateLimiter.test.ts
import { describe, expect, it } from "vitest";
import { RateLimiter } from "./rateLimiter.js";

describe("RateLimiter", () => {
  it("first request is allowed", () => {
    const limiter = new RateLimiter(3, 10);
    expect(limiter.allow("alice", 100)).toBe(true);
  });

  it("requests up to max are allowed", () => {
    const limiter = new RateLimiter(3, 10);
    expect(limiter.allow("alice", 100)).toBe(true);
    expect(limiter.allow("alice", 101)).toBe(true);
    expect(limiter.allow("alice", 102)).toBe(true);
  });

  it("request over the limit is denied", () => {
    const limiter = new RateLimiter(3, 10);
    limiter.allow("alice", 100);
    limiter.allow("alice", 101);
    limiter.allow("alice", 102);

    expect(limiter.allow("alice", 103)).toBe(false);
  });

  it("denials stay sticky inside the same window", () => {
    const limiter = new RateLimiter(3, 10);
    limiter.allow("alice", 100);
    limiter.allow("alice", 101);
    limiter.allow("alice", 102);

    expect(limiter.allow("alice", 103)).toBe(false);
    expect(limiter.allow("alice", 103)).toBe(false);
  });

  it("after the window elapses, a request is allowed", () => {
    const limiter = new RateLimiter(3, 10);
    limiter.allow("alice", 100);
    limiter.allow("alice", 101);
    limiter.allow("alice", 102);

    expect(limiter.allow("alice", 110)).toBe(true);
  });

  it("a new window resets the count", () => {
    const limiter = new RateLimiter(3, 10);
    limiter.allow("alice", 100);
    limiter.allow("alice", 101);
    limiter.allow("alice", 102);

    expect(limiter.allow("alice", 110)).toBe(true);
    expect(limiter.allow("alice", 111)).toBe(true);
    expect(limiter.allow("alice", 112)).toBe(true);
    expect(limiter.allow("alice", 113)).toBe(false);
  });

  it("boundary is strict (>= windowSeconds opens a new window)", () => {
    const limiter = new RateLimiter(3, 10);
    limiter.allow("alice", 100);
    limiter.allow("alice", 101);
    limiter.allow("alice", 102);

    expect(limiter.allow("alice", 109)).toBe(false);
    expect(limiter.allow("alice", 110)).toBe(true);
  });

  it("two clients are independent", () => {
    const limiter = new RateLimiter(3, 10);
    limiter.allow("alice", 100);
    limiter.allow("alice", 101);
    limiter.allow("alice", 102);

    expect(limiter.allow("bob", 102)).toBe(true);
  });

  it("two clients have independent windows", () => {
    const limiter = new RateLimiter(3, 10);
    limiter.allow("alice", 100);
    limiter.allow("bob", 200);

    expect(limiter.allow("alice", 205)).toBe(true);
    expect(limiter.allow("bob", 205)).toBe(true);
  });

  it("a different config works correctly", () => {
    const limiter = new RateLimiter(1, 5);
    expect(limiter.allow("x", 100)).toBe(true);
    expect(limiter.allow("x", 100)).toBe(false);
    expect(limiter.allow("x", 104)).toBe(false);
    expect(limiter.allow("x", 105)).toBe(true);
  });
});

File: challenges/rate-limiter/dart/test/rate_limiter_test.dart
import 'package:rate_limiter_challenge/rate_limiter.dart';
import 'package:test/test.dart';

void main() {
  test('first request is allowed', () {
    final limiter = RateLimiter(3, 10);
    expect(limiter.allow('alice', 100), isTrue);
  });

  test('requests up to the max are allowed', () {
    final limiter = RateLimiter(3, 10);
    expect(limiter.allow('alice', 100), isTrue);
    expect(limiter.allow('alice', 101), isTrue);
    expect(limiter.allow('alice', 102), isTrue);
  });

  test('request over the limit is denied', () {
    final limiter = RateLimiter(3, 10);
    limiter.allow('alice', 100);
    limiter.allow('alice', 101);
    limiter.allow('alice', 102);

    expect(limiter.allow('alice', 103), isFalse);
  });

  test('denials stay sticky inside the same window', () {
    final limiter = RateLimiter(3, 10);
    limiter.allow('alice', 100);
    limiter.allow('alice', 101);
    limiter.allow('alice', 102);

    expect(limiter.allow('alice', 103), isFalse);
    expect(limiter.allow('alice', 103), isFalse);
  });

  test('after the window elapses, a request is allowed', () {
    final limiter = RateLimiter(3, 10);
    limiter.allow('alice', 100);
    limiter.allow('alice', 101);
    limiter.allow('alice', 102);

    expect(limiter.allow('alice', 110), isTrue);
  });

  test('a new window resets the count', () {
    final limiter = RateLimiter(3, 10);
    limiter.allow('alice', 100);
    limiter.allow('alice', 101);
    limiter.allow('alice', 102);

    expect(limiter.allow('alice', 110), isTrue);
    expect(limiter.allow('alice', 111), isTrue);
    expect(limiter.allow('alice', 112), isTrue);
    expect(limiter.allow('alice', 113), isFalse);
  });

  test('boundary is strict (>= windowSeconds opens a new window)', () {
    final limiter = RateLimiter(3, 10);
    limiter.allow('alice', 100);
    limiter.allow('alice', 101);
    limiter.allow('alice', 102);

    expect(limiter.allow('alice', 109), isFalse);
    expect(limiter.allow('alice', 110), isTrue);
  });

  test('two clients are independent', () {
    final limiter = RateLimiter(3, 10);
    limiter.allow('alice', 100);
    limiter.allow('alice', 101);
    limiter.allow('alice', 102);

    expect(limiter.allow('bob', 102), isTrue);
  });

  test('two clients have independent windows', () {
    final limiter = RateLimiter(3, 10);
    limiter.allow('alice', 100);
    limiter.allow('bob', 200);

    expect(limiter.allow('alice', 205), isTrue);
    expect(limiter.allow('bob', 205), isTrue);
  });

  test('a different config works correctly', () {
    final limiter = RateLimiter(1, 5);
    expect(limiter.allow('x', 100), isTrue);
    expect(limiter.allow('x', 100), isFalse);
    expect(limiter.allow('x', 104), isFalse);
    expect(limiter.allow('x', 105), isTrue);
  });
}
